Add auto-heal role and playbooks

jhernand opened the pull request:
This patch adds a new `openshift_autoheal` role that creates the objects
needed by the auto-heal service.

jhernand commented:
@ironcladlou @elad661 @moolitayer @zgalor please review.

ironcladlou commented:
Can you describe how you're testing it? For an end-to-end test this should work:

* Build a new `openshift-ansible` Docker image from this branch
* Use [the release tool](https://github.com/openshift/release/tree/master/cluster/test-deploy) to spin up a new cluster:
  * set `openshift_autoheal_install: true` in `gcp-dev/vars.yml`
  * launch with `make WHAT=dmace-autoheal-test OPENSHIFT_ANSIBLE_IMAGE=openshift-ansible up`

jhernand commented:
I tested it running the installation playbook against an existing cluster that I created with `openshift-ansible` some time ago:

```
$ ansible-playbook -u root -i my.inventory playbooks/openshift-autoheal/config.yml
```

The `my.inventory` file contains the `openshift_autoheal_install=true` variable.

I will re-test using the procedure that you describe.

elad661 commented:
One thing that is missing from both this role and the TPO/cluster-monitoring-operator role is configuring the Alertmanager to connect to the autoheal service.

Since my understanding is that the autoheal service is not meant to be shipped to customers, but the cluster-monitoring-operator *is* meant to be shipped to costumers, I think that the Alertmanager configuration should live in this role.

This might conflict with the wish to give users a declarative way to configure alertmanager by specifying the alertmanager configuration in their inventory file (which was mentioned as something we should consider in the cluster-monitoring-operator role review), so I'm not sure what's the best approach here.

elad661 commented on roles/openshift_autoheal/defaults/main.yml:
```diff
@@ -0,0 +1,7 @@
+---
+openshift_autoheal_awx_ca: ""
```
This variable is not documented in the readme file.

ironcladlou commented:
@elad661 
>One thing that is missing from both this role and the TPO/cluster-monitoring-operator role is configuring the Alertmanager to connect to the autoheal service.

>Since my understanding is that the autoheal service is not meant to be shipped to customers, but the cluster-monitoring-operator is meant to be shipped to costumers, I think that the Alertmanager configuration should live in this role.

>This might conflict with the wish to give users a declarative way to configure alertmanager by specifying the alertmanager configuration in their inventory file (which was mentioned as something we should consider in the cluster-monitoring-operator role review), so I'm not sure what's the best approach here.

For now, the entire Alertmanager config for is a secret maintained by the ops (cluster owner) in a private repo somewhere else. The config can contain a webhook receiver pointing at a previously installed autoheal component.

moolitayer commented on roles/openshift_autoheal/defaults/main.yml:
```diff
@@ -0,0 +1,7 @@
+---
+openshift_autoheal_awx_ca: ""
```
add the new insecure?

elad661 replied:
I think it's probably best not to, as we don't want people running insecure in production. People who want to run insecure in development could manually edit the configuration, or use plain http so there's no illusion of security

ironcladlou commented on roles/openshift_autoheal/tasks/install.yml:
```diff
@@ -0,0 +1,43 @@
+  oc_project:
+    state: present
+    name: "{{ openshift_autoheal_namespace }}"
+    node_selector: ""
```
I'd remove the defaulting here- the misalignment between the default in master-config and the actual node labels is fixed or is being fixed soon anyway.

jhernand replied:
Removed that line.

jhernand commented:
I have updated the pull request to implement the configuration approach that I outlined in a previous comment. Component developers (or anyone wanting to write healing rules) will have to place them into `.yml` files inside the `roles/openshift_autoheal/rules` directory. Each file should contain exactly one rule. The role will merge the basic configuration and all those rule files into one single configuration file, which will then be stored in a configuration map. See the [updated documentation](https://github.com/jhernand/openshift-ansible/tree/add_autoheal_role_and_playbooks/roles/openshift_autoheal#configuration) for details.

@twiest @ironcladlou @moolitayer please review.

jhernand commented:
Thanks for your review @michaelgugino. I have added the parameters to make the image name and version configurable. I have also added the upgrade playbook, but note that it is currently empty because this is the first version of the auto-heal service, so there is no need to upgrade. Please take another look.

michaelgugino commented:
@jhernand 
>  I have also added the upgrade playbook, but note that it is currently empty because this is the first version of the auto-heal service, so there is no need to upgrade. Please take another look.

We support minor upgrades, and all enterprise components should be releasing versions that match the openshift release, IE 3.10.1, 3.10.2.

So, upgrade plays are required on day one to be present.

elad661 commented:
A lot of other components (like our old Prometheus role) do updates simply by running the "install" again with newer image versions. I don't think an empty upgrade playbook should be required - just having the install playbook + configurable image seems enough for me.

@michaelgugino @ironcladlou please correct me if I'm wrong.

moolitayer commented on roles/openshift_autoheal/defaults/main.yml:
```diff
@@ -0,0 +1,21 @@
+---
+# Namespace and name:
+openshift_autoheal_name: autoheal
```
I'm will need to relay on this name not changing in the service monitor definition

moolitayer replied:
and also the namespace being `openshift-autoheal` do we support changing them @jhernand ?

jhernand replied:
This is our call, I guess. Do we want these configurable or not?

jhernand commented on roles/openshift_autoheal/defaults/main.yml:
```diff
@@ -0,0 +1,21 @@
+# Image name:
+openshift_autoheal_image_dict:
+  origin:
+    prefix: "docker.io/openshift/"
```
It isn't needed. It doesn't hurt either. It is used in all the places that use this same mechanism to make the image configurable, so I am preserving it.

michaelgugino replied:
We are moving to full qualified image names, so keeping the docker.io portion is good.

jhernand commented on roles/openshift_autoheal/templates/objects.yml.j2:
```diff
@@ -0,0 +1,118 @@
+    ca.crt: {{ openshift_autoheal_awx_ca | b64encode }}
+
+- apiVersion: v1
+  kind: ConfigMap
```
Note that in this case there are multiple files that need to be merged into one, because the auto-heal service only supports one YAML file. What `oc_configmap` can do is put multiple files inside one secret. That does not solve this problem.

However, I have recently changed the auto-heal service so that it can load multiple configuration files from a directory. Assuming that `oc_configmap` can read multiple files without explicitly enumerating them it could work. I mean something like this:

```yaml
oc_configmap:
  ...
  from_file:
  - rules/*.yml
```

I think that isn't possible right now, but I will check.

ironcladlou replied:
So the way we solved the configuration problem with Alertmanager is to store the entire contents of the config file [as an Ansible variable](https://github.com/openshift/openshift-ansible/blob/master/roles/openshift_cluster_monitoring_operator/defaults/main.yml#L9) which is then consumed by the installer and turned into a secret. That way you can maintain a no-op default and allow the cluster operators (us) to maintain a complete config in an external/private CM system.

I think this configuration system might be simplified if the entire autoheal.yaml (including all AWX connectivity info and rules) were encoded into a single secret file and mounted in for the process to read. Then we can supply an "example" in the default autoheal config variable, and admins (like ops) can easily override.

jhernand commented on roles/openshift_autoheal/README.md:
```diff
@@ -0,0 +1,112 @@
+following variables:
+
+- `openshift_autoheal_deploy`: `true` - install/update. `false` - uninstall.
+  Defaults to `false`.
```
I understand it is a bit confusing. But this follows the usual Ansible way of doing things: you specify what is the state that you desire, and the playbook makes it happen. So if I say that I want the auto-heal service uninstalled (which is the meaning of `openshift_autoheal_deploy=false`) then the playbook should make sure that it is uninstalled.

So, there are no desired *actions* but desired *states*. Maybe the name `..._deploy` is confusing as well, as it is an action. It should probably be `..._deployed`, but looks like other components use `..._deploy` or `..._install`, so I just followed the established practice.

jhernand commented:
@sdodson regarding the use of `oc_apply` versus other `oc_...` tasks, I don't objection to do as  @zgalor suggests. What do you advice?

michaelgugino commented on roles/openshift_autoheal/tasks/upgrade.yml:
```diff
@@ -0,0 +1 @@
+# This is empty on purpose as there is nothing to upgrade.
```
We need to support going from 3.10.0 to 3.10.1 with this task file.  Also, this should support upgrades generally, 3.10 to 3.11.

We require some form of an upgrade play before the role merges.

elad661 replied:
@michaelgugino I think you're wrong here. I don't think `upgrade.yml` should exist *at all* for this role. Like a lot of other roles in openshift-ansible, the proper way to upgrade autoheal would just be changing the version number and then running the `install` tasks again.

A separate upgrade task file should only be needed if there's something special needed to be done on upgrades (like configuration migration). This is not the case here.

sdodson replied:
Yes, it's definitely ideal if upgrading is as simple as calling the install procedures over again. Just call that from your upgrade playbook, as it is now you have an upgrade playbook that's a NOOP. Then if something in the future does require specific logic for upgrades people continue calling the same playbook as before to upgrade and we get that custom migration.

sdodson commented:
> @sdodson regarding the use of oc_apply versus other oc_... tasks, I don't objection to do as @zgalor suggests. What do you advice?

We're trying to move more towards an apply pattern rather than a lot of custom modules for every single object type. So it's fine as is.

jhernand commented:
In the new version of the pull request I addressed some of the comments. I am still ussing `oc_apply`.

@moolitayer I also added the `app: autoheal` label to all the objects, to simplify the creation of the service monitor in the cluster monitoring operator.

zgalor commented on roles/openshift_autoheal/tasks/install.yml:
```diff
@@ -0,0 +1,54 @@
+  when: openshift_autoheal_awx_password is not defined
+
+- name: Create a temporary directory for doing work in the target host
+  command: mktemp -td openshift-autoheal-XXXXXX
```
didn't you say `-t` is deprecated, and `-p` replaces it?

jhernand replied:
Yes, I did, but I wanted to do what you asked :-) . I can replace it with `-p`, no problem.

zgalor replied:
I don't mind which flag you use as long as it does the job :)

jhernand commented:
Replaced `mkdir -td ...` with `mkdir -d -p '' ...`.

moolitayer commented:
> @moolitayer I also added the app: autoheal label to all the objects, to simplify the creation of the service monitor in the cluster monitoring operator.

Thanks @jhernand. Similar components[1] seem to be using `k8s-app: prometheus` I don't know if that distinction is important. @brancz any feedback?

[1]
https://github.com/openshift/cluster-monitoring-operator/blob/3a5bc0c215242515ec70c182b363bf9c755b1060/assets/prometheus-k8s/prometheus-k8s-service-monitor-prometheus.yaml

https://github.com/openshift/cluster-monitoring-operator/blob/3a5bc0c215242515ec70c182b363bf9c755b1060/assets/prometheus-k8s/prometheus-k8s-service-monitor-alertmanager.yaml

https://github.com/openshift/cluster-monitoring-operator/blob/3a5bc0c215242515ec70c182b363bf9c755b1060/assets/prometheus-k8s/prometheus-k8s-service-monitor-node-exporter.yaml

moolitayer commented:
@elad661 yes I meant the label name, sorry if that wasn't clear.

sdodson commented:
> @michaelgugino @sdodson @zgalor so, if I understand correctly, I can just copy the content of main.yml into upgrade.yml because currently, for this component, upgrade is identical to install. Makes sense?

Don't duplicate code, just call install.yml from the upgrade.

brancz commented:
> Thanks @jhernand. Similar components[1] seem to be using k8s-app: prometheus I don't know if that distinction is important. @brancz any feedback?

That labeling scheme was inherited from Tectonic, if we don't like it, I'm happy to change it. I mostly favor consistency here, whether that ends up being k8s-app or component, I don't have a strong opinion - as long as it's consistent.

elad661 commented on roles/openshift_autoheal/tasks/main.yml:
```diff
@@ -0,0 +1,6 @@
+  when: openshift_autoheal_deploy | bool
+
+- include_tasks: uninstall.yml
+  when: not openshift_autoheal_deploy | bool
```
I agree. if `openshift_autoheal_deploy` is false by default this will cause the openshift_autoheal uninstall task file to run on every install of openshift-ansible. It wouldn't do any damage, but it will slightly slow down the install by doing unnecessary stuff.

michaelgugino replied:
The playbook never gets called unless openshift_autoheal_deploy is set to True during the main install of the cluster.

This would only apply if the install/upgrade play was called directly, in which case, defaulting to false would be fine IMO.

ironcladlou commented on roles/openshift_autoheal/tasks/install.yml:
```diff
@@ -0,0 +1,54 @@
+  changed_when: False
+
+- name: Create the list of objects described in the YAML file
+  shell: >
```
@michaelgugino suggested using the `command` module if no stream redirection is used in the command

jhernand replied:
I missed that comment, sorry, it is done now.

ironcladlou commented on roles/openshift_autoheal/templates/objects.yml.j2:
```diff
@@ -0,0 +1,136 @@
+          - --logtostderr
+
+- apiVersion: v1
+  kind: Service
```
Let's [front this with TLS](https://github.com/openshift/openshift-ansible/blob/master/roles/openshift_monitor_availability/files/monitor-app-create.yaml#L76) and put an auth proxy in front of the app process.

ironcladlou commented on roles/openshift_autoheal/templates/objects.yml.j2:
```diff
@@ -0,0 +1,136 @@
+          configMap:
+            name: {{ openshift_autoheal_name }}-config
+        containers:
+        - name: service
```
Should this be secured with the oauth or rbac proxy?

elad661 replied:
I thought we settled on oauth proxy? https://github.com/openshift/autoheal/pull/54

ironcladlou replied:
You're right, it just needs to be used here

elad661 commented on roles/openshift_autoheal/templates/objects.yml.j2:
```diff
@@ -0,0 +1,136 @@
+apiVersion: v1
```
Can we use the same template we have in the autoheal repo itself? https://github.com/openshift/autoheal/blob/master/template.yml

ironcladlou replied:
Sure, although maintaining it in two places is unfortunate, that might be the reality until there's an operator managing this thing. The ansible role will have to be the supported/canonical template in the meantime, I think.

jhernand replied:
Currently this could be an OpenShift template, but it will be problematic in the future if we want to create objects outside of the `openshift-autoheal` namespace, as templates can't create objects outside the namespace where they are applied: the `namespace` field of objects is just overwritten.

So, for example, if in the future we need to create a role binding inside the `kube-system` namespace then we won't be able to use one template, we will need at least two templates, one for each namespace.

On the other hand the list of objects generated with Jinja2 is completely flexible, it can create objects in any namespace.

Note that isn't just theoretical, I already hit this issue during an experiment that I am doing to add an API server that allows us to store things persistently in etcd: I need to add a binding for the `extension-apiserver-authentication-reader` role, which lives in the `kube-system` namespace.

I'd prefer to stick to the Jinja2 template for this reason, but I will change if that is unacceptable. @ironcladlou let me know.

jhernand replied:
By the way, we don't need to wait for the operator to have the template only on one place. When we have this merged we can change the `README.md` of the openshift/authoeal so that it refers to this new role, and then we can remove the template there.

ironcladlou replied:
>Currently this could be an OpenShift template, but it will be problematic in the future if we want to create objects outside of the openshift-autoheal namespace, as templates can't create objects outside the namespace where they are applied: the namespace field of objects is just overwritten.

>So, for example, if in the future we need to create a role binding inside the kube-system namespace then we won't be able to use one template, we will need at least two templates, one for each namespace.

If the role itself is the atomic unit of work, and individual tasks are idempotent, atomicity is preserved across multiple `oc apply` operations. Do you agree that multiple template files could be applied separately against multiple namespaces to achieve the goal in your example?

I only bring this up because my understanding is that all new installable components should have a general focus on divestiture from Ansible insofar as manifests and configuration (at a minimum) are concerned: the architectural direction of the product is to use Go operators to manage these components' installations and upgrades, and in that model, the resources need to be OpenShift Templates (or any other core) resource serializations.

jhernand commented on roles/openshift_autoheal/README.md:
```diff
@@ -0,0 +1,112 @@
+- `openshift_autoheal_deploy`: `true` - install/update. `false` - uninstall.
+  Defaults to `false`.
+
+- `openshift_autoheal_namespace`: Namespace where the components will be
```
OK, will do, thanks @ironcladlou!

jhernand commented on roles/openshift_autoheal/README.md:
```diff
@@ -0,0 +1,112 @@
+- `openshift_autoheal_namespace`: Namespace where the components will be
+  deployed. Defaults to `openshift-autoheal`.
+
+- `openshift_autoheal_name`: Prefix for the names for the objects that compose
```
Done.

jhernand commented on roles/openshift_autoheal/defaults/main.yml:
```diff
@@ -0,0 +1,21 @@
+---
+# Namespace and name:
+openshift_autoheal_name: autoheal
+openshift_autoheal_namespace: openshift-autoheal
```
Done.

jhernand commented:
@ironcladlou regarding your comment about putting everything in a secret: it is doable. We recently did a change in the auto-heal code to support that. I will adjust this pull request to use that.

jhernand commented:
The new version of the pull request addresses all the pending comments, except the use of OpenShift templates instead of Jinja2 templates. In particular it includes the TLS configuration added by @elad661 and it puts all the configuration in the `openshift_autoheal_config` variable as suggested by @ironcladlou.

ironcladlou commented on roles/openshift_autoheal/files/template.yml:
```diff
@@ -0,0 +1,198 @@
+    ports:
+    - name: autoheal
+      port: 443
+      targetPort: 8443
```
I have been using named port references in Service definitions (`public` in this case)... Seems more resilient to change. Just a random observation for future consideration

jhernand replied:
I am open to do that. Can you elaborate? Maybe point me to an example?

ironcladlou replied:
Sure... https://kubernetes.io/docs/concepts/services-networking/service/#defining-a-service

>Note that a Service can map an incoming port to any targetPort. By default the targetPort will be set to the same value as the port field. Perhaps more interesting is that targetPort can be a string, referring to the name of a port in the backend Pods. The actual port number assigned to that name can be different in each backend Pod. This offers a lot of flexibility for deploying and evolving your Services. For example, you can change the port number that pods expose in the next version of your backend software, without breaking clients.

jhernand replied:
Thanks @ironcladlou! @elad661 can you take care of this?

ironcladlou commented on roles/openshift_autoheal/files/template.yml:
```diff
@@ -0,0 +1,198 @@
+    - get
+
+- apiVersion: authorization.openshift.io/v1
+  kind: ClusterRole
```
Are the SA/ClusterRole/ClusterRoleBindings the specific resources you were concerned would need split off into another file? I guess since those are non-namespaced resources we can cheat in this case 😄

jhernand replied:
The objects that I was concerned about aren't needed now. They will be needed when we add custom API and etcd servers to persistently manage the healing rules (and other kinds of objects). Note that the problem will then be to have objects in other namespaces, in particular in the `kube-system` namespace. But as I said that is concert for the future, not now. When the time comes we will solve that, I guess, using two or more templates.

ironcladlou commented:
/lgtm, awesome work. One thing I just noticed is that I was unclear on the config reloading mechanism. Is there one? I know all the CoreOS operators are using a generic configreloader sidecar container.

elad661 commented:
@ironcladlou I think the bot didn't get your lgtm because of the comma. I'll try: /lgtm

Regarding config reloading, this is not implemented yet, but I think the plan is to not use a sidecar container and just do it in-process https://github.com/openshift/autoheal/issues/41

vrutkovs commented:
LGTM needs to be on its own line (or maybe you don't have permissions, not sure)

/lgtm

vrutkovs commented:
/retest

jhernand commented:
@ironcladlou @vrutkovs I think that the GCP upgrade failure isn't related. This is, I think, the cause of the failure:

```
TASK [Wait for instance to respond to SSH] *************************************
Monday 30 April 2018  09:36:43 +0000 (0:00:00.054)       0:00:26.441 ********** 
failed: [localhost] (item={u'status': u'RUNNING', u'network': u'default', u'zone': u'us-east1-c', u'tags': [u'build-image-instance'], u'image': u'rhel-7-20180426-153528', u'disks': [u'prtest-c4a8d93-440-build-image-instance'], u'name': u'prtest-c4a8d93-440-build-image-instance', u'public_ip': u'35.231.179.141', u'private_ip': u'10.142.0.2', u'machine_type': u'n1-standard-1', u'subnetwork': u'default', u'metadata': {}}) => {"changed": false, "elapsed": 121, "item": {"disks": ["prtest-c4a8d93-440-build-image-instance"], "image": "rhel-7-20180426-153528", "machine_type": "n1-standard-1", "metadata": {}, "name": "prtest-c4a8d93-440-build-image-instance", "network": "default", "private_ip": "10.142.0.2", "public_ip": "35.231.179.141", "status": "RUNNING", "subnetwork": "default", "tags": ["build-image-instance"], "zone": "us-east1-c"}, "msg": "Timeout when waiting for 35.231.179.141:22"}
```

Looks like a time-out while waiting for a node to come up. Any suggestion?

vrutkovs commented:
Yeah, seems some GCE issue. Merging manually
